pkg/common/telemetry/telemetry.go

```diff
@@ -0,0 +1,117 @@
+package telemetry
+
+import (
+	"io"
+	"time"
+
+	"github.com/armon/go-metrics"
+)
+
+type SinkConfig struct {
+	Logger      io.Writer
+	ServiceName string
+	StatsdAddr  string
+
+	StopChan <-chan struct{}
+}
+
+// sink carries a small bit of state used for cleanup/shutdown, and
+// implements the Sink interface.
+type sink struct {
+	m *metrics.Metrics
+
+	stopChan    <-chan struct{}
+	inmemCloser func()
+}
+
+// NewSink creates a new sink struct with the appropriate sinks configured.
+func NewSink(c *SinkConfig) Sink {
+	sinks := metrics.FanoutSink{}
+
+	// Always create an in-memory sink
+	interval := 1 * time.Second
+	retention := 1 * time.Hour
+	inmemSink := metrics.NewInmemSink(interval, retention)
+	sinks = append(sinks, inmemSink)
+
+	// Allow the in-memory sink to be signaled, printing stats to the log
+	inmemSignal := metrics.NewInmemSignal(inmemSink, metrics.DefaultSignal, c.Logger)
+
+	if c.StatsdAddr != "" {
+		// Although NewStatsdSink returns an error type, there is no codepath for non-nil error.
+		statsdSink, _ := metrics.NewStatsdSink(c.StatsdAddr)
+		sinks = append(sinks, statsdSink)
+	}
+
+	// Although New returns an error type, there is no codepath for non-nil error.
+	config := metrics.DefaultConfig(c.ServiceName)
+	m, _ := metrics.New(config, sinks)
+
+	t := &sink{
+		m:           m,
+		stopChan:    c.StopChan,
+		inmemCloser: inmemSignal.Stop,
+	}
+
+	go t.wait()
+	return t
+}
+
+func (t *sink) wait() {
```

Comment:
nit: I think we could rename this method, I see it more like a sort of cleanup routine...so maybe it could be called cleanup() or something like that.

Reply:
fixed this, named it `cleanup` instead